### PR TITLE
Optimizing FBL BC and KP scheme

### PR DESCRIPTION
**Pull request by havahol**
Performance optimization of
- the KP scheme
- handling of boundary conditions for FBL

Notebook and script for simulation benchmark for test case paper.

More tests and reference solutions for the FBL scheme.

### CODE REVIEW

File: gpu_ocean/SWESimulators/gpu_kernels/KP07_kernel.cu [hunk 1]
```diff
@@ -293,75 +371,97 @@ __global__ void swe_2D(
...
     //Reconstruct slopes along x and axis
     // The Qx is here dQ/dx*0.5*dx
     minmodSlopeX(Q, Qx, theta_);
-    minmodSlopeY(Q, Qy, theta_);
     __syncthreads();
 
     // Adjust the slopes to avoid negative values at integration points
-    adjustSlopes(Qx, Qy, RHx, RHy, Q);
-    __syncthreads();
+    //adjustSlopes_x(Qx, Hi, Q);
```
Comment (setmar) [inline gpu_ocean/SWESimulators/gpu_kernels/KP07_kernel.cu]: Dead code.

File: gpu_ocean/SWESimulators/gpu_kernels/KP07_kernel.cu [hunk 2]
```diff
@@ -293,75 +371,97 @@ __global__ void swe_2D(
...
     
+    // Reconstruct Q in y-direction while reusing Qx shmem buffer
+    
+    //Reconstruct slopes along x and axis
+    // The Qx is here dQ/dx*0.5*dx
+    minmodSlopeY(Q, Qx, theta_);
+    __syncthreads();
+
+    // Adjust the slopes to avoid negative values at integration points
+    //adjustSlopes_y(Qx, Hi, Q);
```
Comment (setmar) [inline gpu_ocean/SWESimulators/gpu_kernels/KP07_kernel.cu]: Dead code.

File: gpu_ocean/SWESimulators/gpu_kernels/KP07_kernel.cu [hunk 3]
```diff
@@ -384,8 +484,7 @@ __global__ void swe_2D(
             const float hv_b = 0.5f*(hv_a + (Q[2][j][i] + dt_*R3));
             
             //Write to main memory
-            //h_row[ti] = 59.975 + coriolis_f;
```
Comment (setmar) [inline gpu_ocean/SWESimulators/gpu_kernels/KP07_kernel.cu]: Dead code.

File: gpu_ocean/demos/testCasesDemos/run_benchmark.py [hunk 4]
```diff
@@ -89,7 +89,7 @@
 def my_exp(i, j):
         x = dx*i - x_center
         y = dy*j - y_center
-        return np.exp(-10*(x*x/(size*size)+y*y/(size*size))) * (np.sqrt(x**2 + y**2) < size)
+        return np.exp(-10*(x*x/(size*size)+y*y/(size*size))) #* (np.sqrt(x*x + y*y) < size)
```
Comment (setmar) [inline gpu_ocean/demos/testCasesDemos/run_benchmark.py]: Dead code.

File: gpu_ocean/demos/testCasesDemos/Performance_BlockSizeExperiment.ipynb [hunk 5]
```diff
@@ -83,19 +90,331 @@
     "    \n",
     "    \n",
     "#Use the existing file, do not generate new data\n",
-    "#test_filename = \"blocksize_benchmark_run_8.txt\"\n",
+    "test_filename = \"blocksize_benchmark_run_17.txt\"\n",
```
[the remark this answers is not in the document]
Comment (havahol) [reply gpu_ocean/demos/testCasesDemos/Performance_BlockSizeExperiment.ipynb]: This doesn't matter

File: gpu_ocean/SWESimulators/gpu_kernels/KP07_kernel.cu (hunk 1 above)
Comment (havahol) [reply gpu_ocean/SWESimulators/gpu_kernels/KP07_kernel.cu]: Well spotted! This code should of course be included again. It was commented out for the purpose of fair performance benchmarking. Fixed now :+1:
